exoscale/cloudstack
# KVM live migration with non shared storage

Modifications
==========
To achieve the live migration of VMs with non-shared storage, here is a list of things left to be done

Things left
---------------
- [x] Fix hypervisors names to match certificates (https://github.com/exoscale/puppet/pull/75, https://github.com/exoscale/puppet/pull/76, https://github.com/exoscale/puppet/pull/77, https://github.com/exoscale/puppet/pull/78, https://github.com/exoscale/puppet/pull/79, https://github.com/exoscale/puppet/pull/80, https://github.com/exoscale/puppet/pull/81, https://github.com/exoscale/puppet/pull/82, https://github.com/exoscale/puppet/pull/83, https://github.com/exoscale/puppet/pull/84)
- [ ] Fix libgcrypt for encryption support

Configuration changes
-------------------------------
- [ ] Change KVM hypervisor capabilities in the DB (activate `motion_storage`) ✍️
  - [x] preprod
  - [ ] prod, to do after ProdShip
- [x] Change configuration in puppet (see exoscale/puppet#66)
- [ ] Update libvirt on hypervisors to 1.2.21
  - [x] preprod, done manually


Live migration steps
---------------------------
- [x] Unlock CS code to allow `MigrateWithStorage`command for KVM with non shared storage/local storage;
- [x] Check that base image is on the destination host or download it;
- [x] Pre-provisioning of the disk, based on the backing file on the destination host;
- [x] ~~Dump vm XML and change VNC address to the destination's one;~~ _already in CS_
- [x] Send command to libvirt to migrate the vm with flags `(1 << 0)|(1 << 1)|(1 << 2)|(1 << 7)|(1 << 8)|(1 << 12)|(1 << 13)` along with the patched XML description;
  - 0 = VIR_MIGRATE_LIVE - ok
  - 1 = VIR_MIGRATE_PEER2PEER
  - 2 = VIR_MIGRATE_TUNNELLED (to have TLS encryption)
  - 7 = VIR_MIGRATE_NON_SHARED_INC - ok
  - 8 = VIR_MIGRATE_CHANGE_PROTECTION - ok
  - 11 = VIR_MIGRATE_COMPRESSED - ok
  - 12 = VIR_MIGRATE_ABORT_ON_ERROR - ok
  - 13 = VIR_MIGRATE_AUTO_CONVERGE - ok
- [x] Remove vm's disk image file on the source host.


Nice to have
-----------------
When doing manual migration (on the host, through virsh command), the only missing thing is the auto discovery of. CS sees the VM on the new host thanks to the ping command of vms:
- [ ] Volume move of non-shared storage to the destination host;
- [ ] Network traffic should be redirected to the destination host.


Error handling
-------------------
- [x] On failed migration (like a vnc socket error), volume pool id stays with the destination value, state of the volume stays as Migrating. It should be reverted to the source host and to a state Ready;
- [x] Unsure about the IP/network routing update/rollback.
- [x] On migration failure, the VM is stopped
- [x] On migration failure, or success, the volume on the dest or src host is removed


Tests
--------
To be written with @vbernat pytests and added in jenkins as integration tests for cloudstack
- migrate VM while being logged in SSH
- migrate VM with high CPU load
- migrate VM with high IO load
- migrate VM with a few snapshots
  - try revert to previous snapshot
  - try to revert to a snapshot -x (with x > 2)
  - try to revert to 1st snapshot
- migrate Windows VM with a remote connection open
- migrate VM and verify security groups
- try to migrate VM on a host that should not be able to accept the VM (memory exhausted)
- to be continued
- force a failing migration (put a hook?) and validate `cleanup` method

Linked pull requests to commit
=======================
- https://github.com/exoscale/puppet/pull/66
